util.c

```diff
@@ -363,3 +364,302 @@ int UTIL_ensure_path(const char *toDir)
+             "   POWER_PROC=./none\n"
+             "   PARAM=\n\n",
+             name, name, start_time, period, name, path);
+
+   for (curr = points; curr->id; curr++) {
+      int ignore = 0;
+      for (itr = head; itr; itr = itr->next) {
+         if (0 == strcasecmp(itr->name, curr->id)) {
+            ignore = 1;
+            break;
+         }
+      }
+
+      if (mode == 1) {
+         if (!ignore)
+            printf("%s\n", curr->id);
+      }
+      else {
+         if (!ignore) {
+            printf("   <SENSOR>\n");
+            printf("      NAME=%s\n", name);
+            printf("      SENSOR_KEY=%s\n", curr->id);
+            printf("      LOCATION=%s\n", curr->location);
+            printf("      GROUPS=%s\n", curr->group);
+            printf("   </SENSOR>\n");
+         }
+         else {
+            printf("   <IGNORE>\n");
+            printf("      SENSOR_KEY=%s\n", curr->id);
+            printf("   </IGNORE>\n");
+         }
+      }
+   }
+
+   if (header)
+      printf("</SUBPROCESS>\n");
+
+   return 0;
+
+usage:
+   printf("%s [-n <subprocess name>] [-H] [-p <period>] [-s <start time>] [-t <telemetry path] [-x <exclude tlm id>] [-x <>] ...\n",
+      argv[0]);
+   return 0;
+
+}
+
+int UTIL_print_sensor_metadata(struct UTILTelemetryInfo *points,
+   struct UTILEventInfo *events)
+{
+   struct UTILTelemetryInfo *curr;
+   struct UTILBitfieldInfo *bitr;
+   struct UTILEventInfo *eitr;
+   const char *type;
+
+   for (curr = points; curr->id; curr++) {
+      type = "SENSOR";
+      if (curr->computed_by)
+         type = "VSENSOR";
+
+      printf("<%s>\n", type);
+      printf("   KEY=%s\n", curr->id);
+      printf("   LOCATION=%s\n", curr->location);
+      printf("   SUBSYSTEM=%s\n", curr->group);
+      printf("   UNITS=%s\n", curr->units);
+      printf("   DIVISOR=%lf\n", curr->divisor);
+      printf("   OFFSET=%lf\n", curr->offset);
+      printf("   NAME=%s\n", curr->name);
+      printf("   DESCRIPTION=%s\n", curr->desc);
+      if (curr->computed_by)
+      printf("   FUNCTION=%s\n", curr->computed_by);
+
+      if (curr->bitfields) {
+         for (bitr = curr->bitfields; bitr->set_label; bitr++) {
+            printf("   <ENUM>\n");
+            printf("      VALUE=%u\n", bitr->value);
+            printf("      LABEL=%s\n", bitr->set_label);
+            if (bitr->clear_label)
+               printf("      UNSET=%s\n", bitr->clear_label);
+            printf("   </ENUM>\n");
+         }
+      }
+
+      printf("</%s>\n", type);
+   }
+
+   for (eitr = events; eitr->port_name; ++eitr) {
+      printf("<EVENT>\n");
+      printf("   PORT_NAME=%s\n", eitr->port_name);
+      printf("   PORT=%u\n", socket_get_addr_by_name(eitr->port_name));
+      printf("   ID=%u\n", eitr->id);
+      printf("   NAME=%s\n", eitr->name);
+      printf("   DESCRIPTION=%s\n", eitr->desc);
+      printf("</EVENT>\n");
+   }
+
+   return 0;
+}
+
+int UTIL_print_json_telem_dict(struct UTILTelemetryInfo *points,
```

Comment:
If I read this correctly, the output looks like it's in the format of
```
{ ... }
{ ... }
```

which isn't a valid json because it has multiple top level blocks.
Something like this `[{...},{...}]` should parse properly with only minor changes.

Alternatively, something like this
```json
{
  "points": [{},{}],
  "events": [{},{}]
}
```
might be easier to parse by documentation tools by separating events from sensors.

Reply:
You are correct.  There should be an overall top block array.  That needs be be fixed pending the object type resolution.

The current JSON format is a collection of objects, each of which contains a type (sensor, event , etc).  This is a bit more verbose, but somewhat more generic.

We can move the objects to typed arrays, which eliminates the need for the type field (and reduces the overall output size), however it moves the type specifier outside the object, which creates more opportunities for mistakes.

What is the more appropriate tradeoff?